rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -32,491 +33,347 @@
 
 ///=============================================================================
-std::string GenerateToken::liveliness(size_t domain_id)
-{
-  std::string token = "@ros2_lv/" + std::to_string(domain_id) + "/**";
-  return token;
-}
+using Entity = liveliness::Entity;
+using EntityType = liveliness::EntityType;
 
 ///=============================================================================
-/**
- * Generate a liveliness token for the particular entity.
- *
- * The liveliness tokens are in the form:
- *
- * @ros2_lv/<domainid>/<entity>/<namespace>/<nodename>
- *
- * Where:
- *  <domainid> - A number set by the user to "partition" graphs.  Roughly equivalent to the domain ID in DDS.
- *  <entity> - The type of entity.  This can be one of "NN" for a node, "MP" for a publisher, "MS" for a subscription, "SS" for a service server, or "SC" for a service client.
- *  <namespace> - The ROS namespace for this entity.  If the namespace is absolute, this function will add in an _ for later parsing reasons.
- *  <nodename> - The ROS node name for this entity.
- */
-static std::string generate_base_token(
-  const std::string & entity,
-  size_t domain_id,
-  const std::string & namespace_,
-  const std::string & name)
+TopicStats::TopicStats(std::size_t pub_count, std::size_t sub_count)
+: pub_count_(pub_count),
+  sub_count_(sub_count)
 {
-  std::stringstream token_ss;
-  token_ss << "@ros2_lv/" << domain_id << "/" << entity << namespace_;
-  // An empty namespace from rcl will contain "/" but zenoh does not allow keys with "//".
-  // Hence we add an "_" to denote an empty namespace such that splitting the key
-  // will always result in 5 parts.
-  if (namespace_ == "/") {
-    token_ss << "_/";
-  } else {
-    token_ss << "/";
-  }
-  // Finally append node name.
-  token_ss << name;
-  return token_ss.str();
+  // Do nothing.
 }
 
 ///=============================================================================
-std::string GenerateToken::node(
-  size_t domain_id,
-  const std::string & namespace_,
-  const std::string & name)
-{
-  return generate_base_token("NN", domain_id, namespace_, name);
-}
+TopicData::TopicData(
+  liveliness::TopicInfo info,
+  TopicStats stats)
+: info_(std::move(info)),
+  stats_(std::move(stats))
+{}
 
 ///=============================================================================
-std::string GenerateToken::publisher(
-  size_t domain_id,
-  const std::string & node_namespace,
-  const std::string & node_name,
-  const std::string & topic,
-  const std::string & type,
-  const std::string & qos)
+void GraphCache::parse_put(const std::string & keyexpr)
 {
-  std::string token = generate_base_token("MP", domain_id, node_namespace, node_name);
-  token += topic + "/" + type + "/" + qos;
-  return token;
-}
+  std::optional<liveliness::Entity> valid_entity = liveliness::Entity::make(keyexpr);
+  if (!valid_entity.has_value()) {
+    // Error message has already been logged.
+    return;
+  }
```

Comment:
It can be argued how to deal with errors with discovery and whether that should throw and exception or continue the process if the data transport is still working. Maybe something we can iterate on as things get clearer?

Reply:
Opened a ticket to work on in the next iteration https://github.com/ros2/rmw_zenoh/issues/72